### PR TITLE
Optimize cython find_points_in _spheres

### PR DESCRIPTION
**Pull request by lbluque**
## Summary

Optimized cython code in `find_points_in_spheres`, getting ~5x faster runtime.

For example here are the runtimes for searching for all neighbors within a 5A cutoff in a 32 atom FCC (on my laptop with Intel i7 (2.8 GHz) CPU):
- version 2023.5.10: `955 µs ± 35.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)`
- this PR: `115 µs ± 2.69 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)`

## Checklist

Work-in-progress pull requests are encouraged, but please put \[WIP\] in the pull request title.

Before a pull request can be merged, the following items must be checked:

- [X] Doc strings have been added in the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html). Run [pydocstyle](http://www.pydocstyle.org/en/2.1.1/index.html) on your code.
- [X] Type annotations are *highly* encouraged. Run [`mypy path/to/file.py`](https://github.com/python/mypy) to type check your code.
- [X] Tests have been added for any new functionality or bug fixes.
- [X] All linting and tests pass.

Our CI will run all the above checks but it might be more efficient if you already fix most errors before submitting the PR. We highly recommended installing `pre-commit` hooks. Simply Run

```sh
pip install -U pre-commit
pre-commit install
```

in the repo's root directory. Once `pre-commit` has installed `git` hooks, our linters will run before every commit and abort if issues pop up.

### CODE REVIEW

**Comment by shyuep** [comment]
@lbluque NICE! Very useful for all structure manipulations! Merging as teh lint errors are unrelated to PR.

**Comment by janosh** [comment]
@stichri Since you did a great job on https://github.com/materialsproject/pymatgen/pull/2983 checking for memory leaks, would be hugely appreciated if you have the time to take these changes for a spin and let us know if you notice any issues. Always good to have another pair of eyes on performance critical code. 🙏

**Comment by stichri** [comment]
I quickly [tried the new code](https://github.com/materialsproject/pymatgen/files/11578536/memtest.py.txt) and it indeed looks like there is **no issue with leaking memory**:
memtest


Just for curiosity's sake, I also quickly tried to look at it actually using valgrind - to that end, I:
* realized the [suppression file maintained for cpython](https://github.com/python/cpython/blob/3.11/Misc/valgrind-python.supp) really does not seem to suppress anything for me when running a script containing just a single `pass` statement with python3.11
* therefore generated my own suppression file for everything that happens _before_ calling `find_points_in_spheres`  - by commenting out everything from line 15 on in the trial script as linked above and running
   ```
   valgrind --tool=memcheck --leak-check=full --show-leak-kinds=all --error-limit=no --gen-suppressions=all --log-file=all.log python ./memtest.py
   ```
   followed by
   ```
   cat ./all.log | ./parse_valgrind_suppressions.sh > all.supp
   ```
   using [parse_valgrind_suppressions.sh from here](https://wiki.wxwidgets.org/Parse_valgrind_suppressions.sh)
* got to see valgrind's output (_of arguably false positives_) for a single call of `find_points_in_spheres` - by uncommenting line 14 in the trial script as linked above and running
   ```
   valgrind --tool=memcheck --leak-check=full --show-leak-kinds=all --track-origins=yes --suppressions=all.supp python ./memtest.py
   ```
When the new code was compiled with debugging symbols, (as I've done by adding `, extra_compile_args=['-ggdb3,-O0']` on line 173 of pymatgen's setup.py,) I am ultimately left with what seems to be 32 bytes in 2 blocks worth of arguably false positives:
```
==20882== Memcheck, a memory error detector
==20882== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==20882== Using Valgrind-3.21.0 and LibVEX; rerun with -h for copyright info
==20882== Command: python ./memtest.py
==20882==
==20882==
==20882== HEAP SUMMARY:
==20882==     in use at exit: 19,679,587 bytes in 149,555 blocks
==20882==   total heap usage: 1,146,146 allocs, 996,591 frees, 204,187,029 bytes allocated
==20882==
==20882== 16 bytes in 1 blocks are still reachable in loss record 120 of 21,505
==20882==    at 0x484182F: malloc (vg_replace_malloc.c:431)
==20882==    by 0x6259E62: npy_alloc_cache_dim (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62B7AC0: PyArray_NewFromDescr_int (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62BAA92: PyArray_NewLikeArrayWithShape (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62BAD51: PyArray_FromArray (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62BB473: PyArray_FromAny (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62BB5E6: PyArray_CheckFromAny (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x63588A7: array_array (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x4A1ED89: cfunction_vectorcall_FASTCALL_KEYWORDS (methodobject.c:443)
==20882==    by 0x383FF5DF: __Pyx_PyObject_Call (neighbors.c:25731)
==20882==    by 0x383FF5DF: __Pyx__PyObject_CallOneArg (neighbors.c:25770)
==20882==    by 0x3841FF22: __pyx_pf_8pymatgen_12optimization_9neighbors_find_points_in_spheres.isra.0 (neighbors.c:6096)
==20882==    by 0x384239A5: __pyx_pw_8pymatgen_12optimization_9neighbors_1find_points_in_spheres (neighbors.c:3309)
==20882==
==20882== 16 bytes in 1 blocks are still reachable in loss record 121 of 21,505
==20882==    at 0x484182F: malloc (vg_replace_malloc.c:431)
==20882==    by 0x6259E62: npy_alloc_cache_dim (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62B7AC0: PyArray_NewFromDescr_int (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62BAA92: PyArray_NewLikeArrayWithShape (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62BAD51: PyArray_FromArray (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62BB473: PyArray_FromAny (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x62BB5E6: PyArray_CheckFromAny (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x63588A7: array_array (in /home/chrs/venv3.11-pmg-neighbors/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so)
==20882==    by 0x4A1ED89: cfunction_vectorcall_FASTCALL_KEYWORDS (methodobject.c:443)
==20882==    by 0x383FF5DF: __Pyx_PyObject_Call (neighbors.c:25731)
==20882==    by 0x383FF5DF: __Pyx__PyObject_CallOneArg (neighbors.c:25770)
==20882==    by 0x38420208: __pyx_pf_8pymatgen_12optimization_9neighbors_find_points_in_spheres.isra.0 (neighbors.c:6184)
==20882==    by 0x384239A5: __pyx_pw_8pymatgen_12optimization_9neighbors_1find_points_in_spheres (neighbors.c:3309)
==20882==
==20882== LEAK SUMMARY:
==20882==    definitely lost: 0 bytes in 0 blocks
==20882==    indirectly lost: 0 bytes in 0 blocks
==20882==      possibly lost: 0 bytes in 0 blocks
==20882==    still reachable: 32 bytes in 2 blocks
==20882==         suppressed: 19,679,555 bytes in 149,553 blocks
==20882==
==20882== For lists of detected and suppressed errors, rerun with: -s
==20882== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 16378 from 15300)
```
Note the reported offending code lines, e.g. neighbors.c:3309, refer to c code _generated by cython_ which therefore is pretty much unreadable.
So in conclusion, I'm afraid this approach has limited utility without additional tooling (which I wouldn't be aware of) that allows to correlate the offending cython-generated code with the actual pyx-code as written by the user ...

There probably is a much better way that I'm unfortunately not aware of ... 😅

**Comment by janosh** [comment]
Damn, memory leak detection in Cython is a hell of a complex task. Thanks for taking the time and being so thorough!

If it's just 32 bytes out of 20 MiB, arguably we shouldn't sweat it. 😄  I'll just assume it's a timing issue of objects having gone out of scope but the garbage collector not having done a cleanup yet.
